functions/github/webhooks.js

```diff
@@ -0,0 +1,154 @@
+const {
+  getReleaseIdFromBranch,
+  updateCheckRunStatus,
+} = require("../database/database.js");
+const {
+  syncReleaseState,
+} = require("../handlers/handlers.js");
+const {
+  verifySignature,
+} = require("./github.js");
+const {Octokit} = require("@octokit/rest");
+const {error, log} = require("firebase-functions/logger");
+const {defineSecret} = require("firebase-functions/params");
+const GITHUB_WEBHOOK_SECRET = defineSecret("GITHUB_WEBHOOK_SECRET");
+const GITHUB_TOKEN = defineSecret("GITHUB_TOKEN");
+
+/**
+  * Handles a GitHub check run event. This function is called when a check run
+  * is created, completed, or rerequested. It updates the release state if
+  * necessary.
+  *
+  * @param {Object} req - The request object.
+  * @param {Object} res - The response object.
+  * @return {Promise} A promise that resolves when the request is complete.
+  */
+async function githubWebhook(req, res) {
+  if (!verifySignature(req, GITHUB_WEBHOOK_SECRET.value())) {
+    return res.status(401).send("Unauthorized");
+  }
+
+  const payload = req.body;
+  const eventType = req.headers["x-github-event"];
+
+  log("Received GitHub webhook", {payload: payload, eventType: eventType});
+
+  try {
+    if (eventType === "check_run") {
+      await handleCheckRunEvent(payload);
+    } else if (eventType === "pull_request") {
+      await handlePullRequestEvent(payload);
+    }
+  } catch (err) {
+    error("Failed to handle GitHub webhook",
+        {
+          error: err.message,
+          payload: payload,
+          eventType: eventType,
+        },
+    );
+    return res.status(500).send("Internal Server Error");
+  }
+
+  return res.status(200).send("OK");
+}
+
+/**
+  * Handles a GitHub pull request event.
+  *
+  * Updates the release state if the the event implies that the release state
+  * has changed, and needs to be synchronized. As of now, those two events are
+  * opens, and synchronizes (pushes to the pull request branch).
+  *
+  * @param {Object} payload The payload from the GitHub webhook.
+  * @return {Promise<void>}
+  */
+async function handlePullRequestEvent(payload) {
+  if (payload.action === "opened" || payload.action == "synchronize") {
+    const pullRequest = payload.pull_request;
+
+    if (pullRequest && pullRequest.head) {
+      const branchName = pullRequest.head.ref;
+
+      let releaseId;
+      try {
+        releaseId = await getReleaseIdFromBranch(branchName);
+      } catch (err) {
+        error("Error getting release ID from branch name",
+            {error: err.message});
+        return;
+      }
+
+      // If the releaseId exists, then a pull request was opened on a branch
+      // that is being tracked for a release.
+      //
+      // We assume that once a pull request is created on a release branch,
+      // the release configuration has been successfully generated.
+      // If this assumption is wrong, then the release will enter an error
```

Comment:
Q: How do we recover from the error state? And could this webhook be ran again + recover on its own after the release config is ready?

Reply:
To recover from an error state, you'd want to trigger a sync once the issue has been resolved. You can do this through the `modifyReleases` or `refreshRelease` endpoints - they're both going to be able to be called from the dashboard frontend.

As for GitHub webhooks, we can redeliver previous ones, and this could fix the release in some situations. But honestly, it'll likely be easier to just refresh the release on the front end.